### ISSUE TITLE
Making Nancy run on .NET Core

### ISSUE THREAD

**Issue by jchannon** [NancyFx/Nancy#1959]
Obviously we want to make Nancy run on .NET Core so we'll use this issue as the parent issue.

- [ ] Restructure repository to aid project.json port (#2154)
- [ ] Convert production projects to project.json (#2155)
- [ ] Convert demo projects to project.json
- [ ] Convert test projects to project.json (this possibly involves upgrading xUnit to 2.1)
- [ ] Add `dotnet5.3` (`netstandard1.2`) targets to run on .NET core (and beyond!)
- [ ] Revamp build infrastructure to build on .NET, Mono and .NET core, on both Linux and Windows (Travis and AppVeyor?)
- [ ] :fire: all Ruby-related stuff
- [ ] :fire: old .csprojs, packages.config, nuspec etc.

Please add items to this list or ask @NancyFx/most-valued-minions @NancyFx/owners to append it

**Comment by awec** [comment]
[] Get Nancy compiling on dnx core (removing anything that causes build break for now)

**Comment by awec** [comment]
- [ ] Get Nancy compiling on dnx core (removing anything that causes build break for now)

**Comment by awec** [comment]
I ran Nancy.dll through the portability analyzer, I've uploaded the spreadsheet here:

http://1drv.ms/1IokAgu

Lists all the references that are not supported in aspnet5 and an alternative in some cases. Most of the System.Type ones are for TinyIOC which we can ignore as there is already a coreclr version of that available.

**Comment by phillip-haydon** [comment]
@awec file doesn't exist YO!

**Comment by tugberkugurlu** [comment]
One other thing to consider would be to sit on top of [the new routing system](https://github.com/aspnet/Routing) rather having a custom one. For example, [MVC 6 today hooks into this routing system](https://github.com/aspnet/Mvc/blob/fc2019c973e84c8b7faa99793cba56412e032c70/src/Microsoft.AspNet.Mvc.Core/BuilderExtensions.cs#L55-L78).

**Comment by phillip-haydon** [comment]
@tugberkugurlu you mean pass the routes from a module back to the middleware? Or move the routes out of the modules all together and use the middleware to point to routes?

**Comment by grumpydev** [comment]
We won't be replacing our routing, if people want to use the ASPNet stuff to do mapping that's fine, but that's no different to how we currently use OWIN (and without any of the IApplicationBuilder nonsense - bleh :))

**Comment by EliotJones** [comment]
I gave a go at cutting down some of the build errors but messed git up a bit in the process: https://github.com/EliotJones/NancyTest/tree/feature/core50 . For me this is building against DNXCore50 with only 200 errors but it might just be my environment, sorry if it's still broken for everyone else

**Comment by jchannon** [comment]
@awec where's that coreclr version of tinyioc you mention?

**Comment by janus007** [comment]
So... does this mean I can use this branch in my DNX 451 project?

**Comment by jchannon** [comment]
You can try but it's not complete

**Comment by masaeedu** [comment]
@janus007 I've been using the latest `Nancy` and `Nancy.Owin`, along with the `UseOwin` extension method available in the [Microsoft.AspNet.Owin](https://www.nuget.org/packages/Microsoft.AspNet.Owin) package, and things seem to have been running smoothly so far. I'll re-emphasize that this is DNX **451**, i.e. running on the full CLR. Here's what my Startup class looks like:

    class Startup
    {
        public void Configure(IApplicationBuilder app)
        {
            var nancy = NancyMiddleware.UseNancy(NancyConfiguration);

            // Give Nancy first crack at serving the request
            app.UseOwin(adder => adder.Invoke(nancy));

            // Add any other middleware you want
            app.UseDefaultFiles();
            app.UseStaticFiles();
        }

        private void NancyConfiguration(NancyOptions opts)
        {
            // Configure Nancy however you want
        }
    }

**Comment by khellang** [comment]
@masaeedu You know that we provide an extension method for the `AddMiddleware` delegate in ASP.NET 5, so you can do

```csharp
// Either
app.UseOwin(owin => owin.UseNancy(NancyConfiguration));

// Or
app.UseOwin().UseNancy(NancyConfiguration);
```

**Comment by masaeedu** [comment]
@khellang Ah, I didn't realize there was a `UseNancy` for the "MidFunc eater" delegate. Apologies for any misinformation.

**Comment by tpluscode** [comment]
Hi. I've just created a bare-bone Nancy app in DNX 4.5.1 with a module in a separate class library. It turns out that Nancy cannot discover modules from that library, because the reference is actually not added until some code actually referenced. The typical way Nancy discovers stuff by inspecting all assemblies in `bin` also doesn't work, because apparently there is no bin folder, or is there?

Is there any way other than actually referencing some code explicitly? This breaks the feature I liked about Nancy, that I could just drop a dll in bin and it just works.

**Comment by khellang** [comment]
@tpluscode This is the same for all Nancy projects - DNX or not. Nancy won't scan assemblies not referencing Nancy.dll by default.
If you don't reference the dll or not using any Nancy code in your library at compile-time (which makes the compiler strip the reference), it won't be picked up by default.
We even added the `[assembly: IncludeInNancyAssemblyScanning]` attribute just for this scenario.

Also, there's no hardcoded path that Nancy scans. It looks at `AppDomain.CurrentDomain.SetupInformation.ApplicationBase` and `AppDomain.CurrentDomain.SetupInformation.PrivateBinPath`, so AFAIK this works on DNX451 as well :smile:

**Comment by tpluscode** [comment]
@khellang AFAICT `[assembly: IncludeInNancyAssemblyScanning]` helps when a class library doesn't reference Nancy (or that assembly is removed at compilation). Is that right?

In my case the Nancy hosted with OWIN in a ASP.NET 5 app doesn't discover my module. `ApplicationBase` is the DNX path, so that won't help and it turns out that the `PrivateBinPath` is an empty string. Any ideas?

Also, [here's a repro](https://github.com/tpluscode/nancy-dnx/).

**Comment by khellang** [comment]
Hmm. When I think about it, I doubt anyone's actually tested this scenario. It's probably just been tested with modules in the same assembly as the host. Until we get up and running with DNX's `ILibraryManager`, I think we'll have to find a workaround here. I think you have to inject `IApplicationEnvironment` into your `Startup` and set `PrivateBinPath` to its `ApplicationBasePath` or something.

**Comment by khellang** [comment]
Some context; https://github.com/aspnet/dnx/issues/1253

**Comment by Alxandr** [comment]
@khellang getting it to run with `ILibraryManager` should be really easy. Here's the entire MVC's implementation (really few LOCs): https://github.com/aspnet/Mvc/blob/dev/src/Microsoft.AspNet.Mvc.Core/Infrastructure/DefaultAssemblyProvider.cs

**Comment by khellang** [comment]
Absolutely. But there are some yaks that needs some shaving before we get to that point;
We're aiming to empty our PR/issue queue, ship 1.4, then start (continue) on 2.0, which will move to project.json, drop 4.0 etc.

**Comment by khellang** [comment]
One part of the problem is that Nancy's assembly scanning is static today, while `ILibraryManager` is instance based, so we have to adjust the assembly scanning a bit. See https://github.com/NancyFx/Nancy/pull/1846 for context.

**Comment by danfma** [comment]
Hey guys, how is the work on this feature?

**Comment by khellang** [comment]
@danfma We're working on some stuff to unblock .NET core. `net451` etc. should work fine.

**Comment by danfma** [comment]
Can I help with anything? I don't have much time but I can try to help with something.

**Comment by khellang** [comment]
The problem is that there's not enough work to do it concurrently. We'll just have to let @thecodejunkie finish #1846 after #2000 is done :smile: Thanks for offering your help! :+1:

**Comment by danfma** [comment]
Ok!

**Comment by jchannon** [comment]
UPDATE : I ditched my branch as it got too big and we plan to move to dnx in smaller stages. At time of writing we already have a couple of PRs to change the repo structure and have csproj and project.json side by side in the short term

**Comment by RTodorov** [comment]
Hi guys,

Is someone still working on this? Can we help somehow?

http://www.ageofascent.com/asp-net-core-exeeds-1-15-million-requests-12-6-gbps/

Looking forward to see Nancy doing it! ;)

**Comment by Yantrio** [comment]
looks like this issue is a bit old, we just got nancy compiling on the coreclr today (about 3 hours ago) and you can see our efforts over on the coreclr branch (issue #2220 tracked most of the issues)

@jchannon can we close this now?

**Comment by jchannon** [comment]
This is kind of a duplicate of #2220 so closing but if anyone is following check these out:

https://twitter.com/NancyFx/status/702825092418543616
https://twitter.com/jchannon/status/702835712224325632

**Comment by RTodorov** [comment]
Wow, that is great!! Nice timing! Are you planing to release a version anytime soon?

**Comment by jchannon** [comment]
Not at the moment. We plan to do a 2.0-alpha soon from master branch, then
once we have all of nancy projects eg/nancy testing, authentication etc
compiling I imagine we may do an alpha/beta release of that but this purely
guess work at the moment

**Comment by peterblazejewicz** [comment]
@jchannon 
What is the status of `dotnet core` at the moment? I'm trying to port `Nancy.Demo.Hosting.Kestrel/Person.cs` for RTM on generator-aspnet but I have problems (OS X/ dotent 1.0.0-preview2-003121)
Thanks!

**Comment by jchannon** [comment]
dotnet core is compiling on the master branch, we have no nuget package out
yet, we are working on getting our build script over to cake, once that is
done we can release a nuget package

**Comment by peterblazejewicz** [comment]
Thanks!

**Comment by a23o** [comment]
Hi guys!
How is the cake build going? Do you have a roadmap for the nuget package? 
I appreciate your hard work.

Thx

**Comment by cemremengu** [comment]
@a23o already done :) checkout the latest barneyrubble release

**Comment by a23o** [comment]
Awesome!! Thank you.

**Comment by mrgrifon** [comment]
Thank you very much for the great work you.
I was able to run Nancy under .NET Core
But I am having a problems with Nancy.Hosting.Self 

Package Nancy.Hosting.Self 2.0.0-barneyrubble is not compatible with netcoreapp1.0 (.NETCoreApp,Version=v1.0). Package Nancy.Hosting.Self 2.0.0-barneyrubble supports: net452 (.NETFramework,Version=v4.5.2)

As I can see from sources you already updated this package.

![hostingself-updated](https://cloud.githubusercontent.com/assets/927700/17312359/f2359da8-5806-11e6-8f79-96359e4ad947.png)

I am running .NET Core on Linux Ubuntu 14.04 LTS
Could you help me?

**Comment by phillip-haydon** [comment]
@mrgrifon You should use Kesteral hosting, please refer to https://github.com/NancyFx/Nancy/tree/master/samples/Nancy.Demo.Hosting.Kestrel

**Comment by erinxocon** [comment]
Will self hosting be implemented for .NET core, or is the plan to just use Kestral?

**Comment by khellang** [comment]
Kestrel *is* self-hosting. It's the default model for ASP.NET Core apps. Either using Kestrel, or WebListener.

**Comment by mmaguigan** [comment]
UseOwin seems to be missing from Kestrel in .NET Core...

**Comment by khellang** [comment]
`UseOwin` has never been a part of Kestrel? You need to install `Microsoft.AspNetCore.Owin` to get the `UseOwin` extension method.

**Comment by mmaguigan** [comment]
khellang, it's installed and is no longer present with .NET Core 2.0 preview 1.

**Comment by reddy6ue** [comment]
Thoroughly confused right now. So, is the netcore 2.0 port happening?

**Comment by fizxmike** [comment]
The thing I (probably we) loved about Nancy.Hosting.Self is the MINIMAL boilerplate... That kestrel demo project is a nightmare for anyone getting started with dotnet core and Nancy IMHO. I don't mind switching to kestrel, just give us a minimalist boilerplate demo PLEASE!?

Right now, OWIN looks like the way to go maybe since someone posted a good example and people have indicated it works.

**Comment by fizxmike** [comment]
Alright... here is a little gift to anyone who wants to do this but couldn't find a good example

#Nancy on Dotnet Core Minimal Example (OWIN + Kestrel)

## HelloWorld.csproj
```
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <AssemblyName>HelloWorld</AssemblyName>
    <OutputType>Exe</OutputType>
    <TargetFramework>netcoreapp2.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="Microsoft.AspNetCore.Hosting" Version="2.0.2" />
    <PackageReference Include="Microsoft.AspNetCore.Owin" Version="2.0.2" />
    <PackageReference Include="Microsoft.AspNetCore.Server.Kestrel" Version="2.0.2" />
    <PackageReference Include="Nancy" Version="2.0.0-clinteastwood" />
    <PackageReference Include="Nancy.Owin" Version="2.0.0-clinteastwood" />
  </ItemGroup>

</Project>
```
## Program.cs
```
using System;
using Microsoft.AspNetCore.Hosting;

namespace HelloWorld
{
    public class Program
    {
        public static void Main(string[] args)
        {
            var host = new WebHostBuilder()
                .UseKestrel()
                .UseStartup<Startup>()
                .Build();

            host.Run();
        }
    }
}
```

##Startup.cs
```
using Nancy.Owin;
using Microsoft.AspNetCore.Owin;
using Microsoft.AspNetCore.Builder;

namespace HelloWorld
{
    public class Startup
    {
        public void Configure(IApplicationBuilder app)
        {
            app.UseOwin().UseNancy(NancyConfiguration);
        }

        private void NancyConfiguration(NancyOptions opts)
        {
            // Configure Nancy however you want
        }
    }
}

```

## DefaultModule.cs
```
using Nancy;

namespace HelloWorld
{
     public sealed class DefaultModule : NancyModule
    {
        public Router()
        { 
            Get("/", _ => "Hello (dotnet core) World!");
        }
    }
}
```
